sympy/combinatorics/free_groups.py

```diff
@@ -344,27 +325,70 @@ def center(self):
 ############################################################################
 
 
-class FreeGroupElement(CantSympify, DefaultPrinting, tuple):
+class FreeGroupElement(Basic):
     """Used to create elements of FreeGroup. It can not be used directly to
     create a free group element. It is called by the `dtype` method of the
     `FreeGroup` class.
 
+    Equality between free group elements
+    ====================================
+
+    Two associative words are equal if they are words over the
+    same alphabet and if they are sequences of the same letters.
+    This is equivalent to saying that the external representations
+    of the words are equal.
+    There is no "universal" empty word, every alphabet has its own
+    empty word.
+
+    >>> from sympy.combinatorics.free_groups import free_group
+    >>> f, swapnil0, swapnil1 = free_group("swapnil0 swapnil1")
+    >>> f
+    <free group on the generators (swapnil0, swapnil1)>
+    >>> g, swap0, swap1 = free_group("swap0 swap1")
+    >>> g
+    <free group on the generators (swap0, swap1)>
+
+    >>> swapnil0 == swapnil1
+    False
+    >>> swapnil0*swapnil1 == swapnil1/swapnil1*swapnil0*swapnil1
+    True
+    >>> swapnil0*swapnil1 == swapnil1*swapnil0
+    False
+    >>> swapnil1**0 == swap0**0
+    False
     """
     is_assoc_word = True
 
-    def new(self, init):
-        return self.__class__(init)
+    _array_form = None
+    _group = None
+
+    def __new__(cls, array_form, group):
+        if isinstance(array_form, Tuple):
+            array_form = tuple(tuple(i) for i in array_form)
 
-    _hash = None
+        if not isinstance(array_form, tuple):
+            raise ValueError("{} must be a tuple.".format(array_form))
+        if not isinstance(group, FreeGroup):
+            raise ValueError("{} must be a free group.".format(group))
 
-    def __hash__(self):
-        _hash = self._hash
-        if _hash is None:
-            self._hash = _hash = hash((self.group, frozenset(tuple(self))))
-        return _hash
+        return cls._new(array_form, group)
+
+    @classmethod
+    def _new(cls, array_form, group):
+        obj = super().__new__(cls)
+        obj._array_form = array_form
+        obj._group = group
```

Comment:
Should these not be passed to `__new__` and stored in the `_args`?

Reply:
I have seen that there was a noticable slowdown while running the combinatorics test suite 
`30.09 seconds => 61.53 seconds` when the tuple is sympified.
While making the node atomic and the `args` lazily generated had been reasonable `28.42 seconds => 30.09 seconds`.

So I'm looking for changes that does not introduce some slowdowns.
What do you mean by things should be stored in `_args`. I see it also needs to be sympified.

Reply:
I don't know whether these should be made to be `Basic` (it looks like they are deliberately not Basic just like the poly domain elements).

If these are Basic though then the args should normally be passed to `Basic.__new__` so that they can be stored in `_args` and the hash can be computed correctly.

Reply:
I think that `FreeGroupElement` had grown up quite too much containing not only the crude arithmetics, but also lots of unimportant computations as its methods.

But I may think of some other ideas like boxing and unboxing.

Reply:
One other problem is that it's not possible to use these stuff as `Matrix`, `Set` on top of the nonsympifyable objects. 
So there should be some respective sympifyable element as well as the group itself.
I think that sympy crude number types like `Integer` are atomic because it can't contain `Integer` itself in its args. So I've used the similar approach.